File: backend/app/models/investment.rb
class Investment < ApplicationRecord
  belongs_to :funder
  belongs_to :project

  has_many :investment_subgeographics, dependent: :destroy
  has_many :subgeographics, through: :investment_subgeographics
  has_many :subgeographic_ancestors, through: :subgeographics, source: :subgeographic_ancestors

  validates :demographics, array_inclusion: {in: Demographic::TYPES, allow_blank: true}
  validates :funding_type, inclusion: {in: FundingType::TYPES, allow_blank: true}
  validates :capital_type, inclusion: {in: CapitalType::TYPES, allow_blank: true}, presence: true
  validates :areas, array_inclusion: {in: Area::TYPES, allow_blank: true}, presence: true
  validates :grant_duration, inclusion: {in: GrantDuration::TYPES, allow_blank: true}, presence: true
  validates :privacy, inclusion: {in: InvestmentPrivacy::TYPES, allow_blank: true}, presence: true

  validates :number_of_grant_years, presence: true, if: -> { grant_duration == "multi_year" }
  validates :funding_type, presence: true, if: -> { capital_type == "grants" }
  validates :year_invested, :initial_funded_year, numericality: {only_integer: true, greater_than: 1980, less_than: 2100}
  validates :amount, numericality: {greater_than: 0}

  scope :can_show_aggregated_amount, -> { where privacy: %w[all aggregate_amount_funded] }
  scope :can_be_shown_without_amount, -> { where privacy: %w[all aggregate_amount_funded amount_funded_visible_only_to_members amount_funded_visible_only_to_staff] }
end

File: client/src/types/api/investment.ts
import { CapitalType } from '@/types/api/capital-type';
import { Demographic } from '@/types/api/demographic';
import { FundingType } from '@/types/api/funding-type';
import { DurationGranInvestment } from '@/types/api/grant-duration';
import { Privacy } from '@/types/api/privacy';
import { Project } from '@/types/api/project';

export interface Investment {
  id: string;
  amount: string;
  year_invested: number;
  initial_funded_year: number;
  funding_type: FundingType;
  funding_type_other: string;
  areas: string[];
  areas_other: string;
  grant_duration: DurationGranInvestment;
  number_of_grant_years: number;
  demographics: Demographic[];
  demographics_other: string | undefined;
  capital_type: CapitalType;
  capital_type_other: string | undefined;
  submitting_organization_contact_name: string | undefined;
  privacy: Privacy;
  updated_at: string;
  created_at: string;
  project: Project;
  subgeographic_ancestors: {
    type: 'subgeographic';
    id: string;
  }[];
  subgeographics: {
    type: 'subgeographic';
    id: string;
  }[];
}

export type InifiniteInvestments = {
  data: Investment[];
  meta: {
    page: number;
    pages: number;
    total: number;
    to: number;
  };
};
